File: LC-Java/src/Tree/_124_maxPathSum.java
package Tree;

import DataStructure.TreeNode;

public class _124_maxPathSum {

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left=new TreeNode(2);
        root.right=new TreeNode(3);
        root.left.left=new TreeNode(4);
        root.left.right=new TreeNode(5);
        root.right.right=new TreeNode(7);
        root.right.left=new TreeNode(6);
        root.left.left.left=new TreeNode(8);
        root.left.left.right=new TreeNode(9);

        int max = new _124_maxPathSum().maxPathSum(root);
        System.out.println(max);
    }

    int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        helper(root);
        return max;
    }

    private int helper(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = Math.max(helper(root.left), 0);
        int right = Math.max(helper(root.right), 0);

        max = Math.max(max, left + right + root.val);

        return root.val+Math.max(left,right);
    }
}

File: TopKFromLCTags/InterviewPrepare/124-maxPathSum.go
package interviewprepare

import "math"

func maxPathSum(root *TreeNode) int {
	maxSum := math.MinInt32
	maxTree(root, &maxSum)
	return maxSum
}
func maxTree(node *TreeNode, maxSum *int) int {
	if node == nil {
		return 0
	}
	left := maxMaxPathSum(maxTree(node.Left, maxSum), 0)
	right := maxMaxPathSum(maxTree(node.Right, maxSum), 0)
	*maxSum = maxMaxPathSum(*maxSum, node.Val+left+right)
	return node.Val + maxMaxPathSum(left, right)
}
func maxMaxPathSum(i, j int) int {
	if i >= j {
		return i
	}
	return j
}
